yarn-project/prover-client/src/proving_broker/proving_broker.ts

```diff
@@ -0,0 +1,341 @@
+    const result = this.resultsCache.get(id);
+    if (!result) {
+      // no result yet, check if we know the item
+      const item = this.jobsCache.get(id);
+
+      if (!item) {
+        this.logger.warn(`Proving job id=${id} not found`);
+        return Promise.resolve({ status: 'not-found' });
+      }
+
+      return Promise.resolve({ status: this.inProgress.has(id) ? 'in-progress' : 'in-queue' });
+    } else if ('value' in result) {
+      return Promise.resolve({ status: 'resolved', value: result.value });
+    } else {
+      return Promise.resolve({ status: 'rejected', error: result.error });
+    }
+  }
+
+  // eslint-disable-next-line require-await
+  async getProvingJob<T extends ProvingRequestType[]>(
+    filter: ProvingJobFilter<T> = {},
+  ): Promise<V2ProvingJob | undefined> {
+    const allowedProofs: ProvingRequestType[] = filter.allowList
+      ? [...filter.allowList]
+      : Object.values(ProvingRequestType).filter((x): x is ProvingRequestType => typeof x === 'number');
+    allowedProofs.sort(proofTypeComparator);
+
+    for (const proofType of allowedProofs) {
+      const queue = this.queues[proofType];
+      const item = queue.getImmediate();
+      if (item) {
+        this.inProgress.set(item.id, {
+          id: item.id,
+          startedAt: this.timeSource(),
+          lastUpdatedAt: this.timeSource(),
+        });
+
+        return item;
+      }
+    }
+
+    return undefined;
+  }
+
+  async reportProvingJobError<T extends ProvingRequestType>(
+    id: V2ProvingJobId<T>,
+    err: Error,
+    retry = false,
+  ): Promise<void> {
+    const info = this.inProgress.get(id);
+    const item = this.jobsCache.get(id);
+    const retries = this.retries.get(id) ?? 0;
+
+    if (!item) {
+      this.logger.warn(`Proving job id=${id} not found`);
+      return;
+    }
+
+    if (!info) {
+      this.logger.warn(`Proving job id=${id} not in the in-progress set`);
+    } else {
+      this.inProgress.delete(id);
+    }
+
+    if (retry && retries + 1 < this.maxRetries) {
+      this.logger.info(`Retrying proving job id=${id} retry=${retries + 1}`);
+      this.retries.set(id, retries + 1);
+      this.enqueueJobInternal(item);
+      return;
+    }
+
+    this.logger.debug(`Marking proving job id=${id} totalAttempts=${retries + 1} as failed`);
+    await this.database.setProvingJobError(id, err);
+    this.resultsCache.set(id, { error: String(err) });
+  }
+
+  reportProvingJobProgress<T extends ProvingRequestType, F extends ProvingRequestType[]>(
+    id: V2ProvingJobId<T>,
+    filter?: ProvingJobFilter<F>,
+  ): Promise<V2ProvingJob | undefined> {
+    const metadata = this.inProgress.get(id);
+    if (metadata) {
+      metadata.lastUpdatedAt = this.timeSource();
+      return Promise.resolve(undefined);
+    } else if (filter) {
+      this.logger.warn(`Proving job id=${id} not found in the in-progress set. Sending new one`);
+      return this.getProvingJob(filter);
```

Comment:
I think it's because this design can result in more than 1 agent proving a job. No information is persisted relating to a job currently being proven. This is to drastically reduce DB writes. The result is that after a restart, the broker doesn't know which jobs are in progress and could re-issue them. 

I think the idea here is that if I was proving job 'x' and it is not longer in progress here then it must have been completed by somebody else and I get given a new job. Is that right @alexghr?

However, I would have thought we would want some more logic here. If `metadata` contained an id of who has the job then we can see if somebody else has the same job _before_ it's finished couldn't we?

Reply:
> I think the idea here is that if I was proving job 'x' and it is not longer in progress here then it must have been completed by somebody else and I get given a new job. Is that right @alexghr?

Correct 👍 

> However, I would have thought we would want some more logic here. If metadata contained an id of who has the job then we can see if somebody else has the same job before it's finished couldn't we?

You're right, this is currently unhandled so two agents can work the same job in parallel (if the broker restarts). Will fix.

Reply:
This is now correctly handled:
- broker crashes and restarts losing in-memory data
- old agent reports progress => new broker instance updates job as in progress
- new agent reports progress => it receives a new job

The broker also now correctly handles the case where after a crash it doesn't return in-progress jobs